### PR TITLE
plugin/pkg/auth: add openid authenticator.

### PR DESCRIPTION
**Pull request by yifan-gu**
Add apiserver side code to verify the token get from the open id provider. Assuming `jwks_uri` is available for now.

### CODE REVIEW

**Comment by yifan-gu** [comment]
The last commits implements a token authenticator for the openid protocol. This will require `api-server` to have two more cmdline flags: `--openid-endpoint` and `--openid-client-id`.

Basically if the authenticator is enabled, it will verify and decode the bearer token in the request. <del>After the verification, the authenticator will cache the token until it expires.</del>

For now, only providers that have implemented `jwks_uri` are supported.

Tested manually by `curl -v --insecure https://0.0.0.0:6443/api/v1/nodes -H "Authorization: Bearer $TOKEN`
It works fine, will add httptest later.

/cc @bcwaldon @bobbyrullo @philips

**Comment by yifan-gu** [comment]
Fix https://github.com/coreos-inc/tectonic/issues/34

**Comment by yifan-gu** [comment]
@bcwaldon @bobbyrullo @philips @sym3tri Tests added, ready for some review :)

File: plugin/pkg/auth/authenticator/token/openid/openid.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/auth/user"
+	"github.com/coreos/go-oidc/jose"
+	"github.com/coreos/go-oidc/oidc"
+	"github.com/golang/glog"
+)
+
+var (
+	maxRetrial   = 5
```
Comment (philips) [inline plugin/pkg/auth/authenticator/token/openid/openid.go]: maxRetries is a better name here

**Comment by bobbyrullo** [comment]
LGTM

**Comment by yifan-gu** [comment]
Cool, I am going to PR the upstream!

**Comment by yifan-gu** [comment]
Pushed into [oidc branch](https://github.com/coreos/kubernetes/tree/oidc) /cc @bcwaldon

**Pull request closed without merging**